core/src/cubos/core/data/ser/json.cpp

```diff
@@ -0,0 +1,97 @@
+#include <cubos/core/log.hpp>
+
+/// [JSON serializer]
+#include <cubos/core/data/ser/json.hpp>
+
+using cubos::core::data::JSONSerializer;
+using cubos::core::reflection::Type;
+
+auto jsonVal = nlohmann::json::object(); ///< The current JSON value that is being serialized.
+
+/// [MACRO to add hooks for primitive types]
+#define AUTO_HOOK(type, ...)                                                                                           \
+    this->hook<type>([](const type& value) {                                                                           \
+        jsonVal = value;                                                                                               \
+        return true;                                                                                                   \
+    })
+
+/// [Setting up hooks]
+#include <cubos/core/reflection/external/primitives.hpp>
+#include <cubos/core/reflection/external/string.hpp>
+
+using cubos::core::reflection::reflect;
+
+JSONSerializer::JSONSerializer()
+{
+    AUTO_HOOK(bool);
+    AUTO_HOOK(char);
+    AUTO_HOOK(int8_t);
+    AUTO_HOOK(int16_t);
+    AUTO_HOOK(int32_t);
+    AUTO_HOOK(int64_t);
+    AUTO_HOOK(uint8_t);
+    AUTO_HOOK(uint16_t);
+    AUTO_HOOK(uint32_t);
+    AUTO_HOOK(uint64_t);
+    AUTO_HOOK(float);
+    AUTO_HOOK(double);
+    AUTO_HOOK(std::string);
+}
+/// [Setting up hooks]
```

Comment:
```suggestion
```
These comments are only used in samples to create snippets which can be referenced from the markdown pages.